### PR TITLE
Prevent mixed case env vars from crashing processes like worker

### PR DESCRIPTION
**Pull request by haard**
An environment variable that almost looks like an Airflow var but is not (e.g `AIRFLOW__SOME__brokenthing`) leads to a NPE in `airflow/configuration.py`. In addition, when using the Celery Executor and triggering this exception, the context is silenced in the log message (using `.error` instead of `.exception`) which makes troubleshooting needlessly hard.

This patch ignores but logs env vars that have trailing lowercase part, and uses `log.exception` when running a Celery task fails.

### CODE REVIEW

**Comment by XD-DENG** [comment]
Hi @haard I assume you meant the cases where the env var cannot be found properly if the *case* is not given correctly?

File: tests/core/test_configuration.py
Comment (ashb) [inline tests/core/test_configuration.py:89]: This is not the behaviour I would expect. It should instead still be present.

There are a few plugins out there that add their own config sections, and that is something we want to support.

**Review by ashb — DISMISSED** [review]

**Comment by haard** [comment]
> 
> 
> Hi @haard I assume you meant the cases where the env var cannot be found properly if the _case_ is not given correctly?

Yes, sorry. Edited the description, I had started writing it before I pinned down the issue properly.

File: tests/core/test_configuration.py
Comment (haard) [reply tests/core/test_configuration.py:89]: > This is not the behaviour I would expect. It should instead still be present.
> 
> There are a few plugins out there that add their own config sections, and that is something we want to support.

If they do and do not enforce ALL_UPPER_CASE they will break if using `as_dict()` with the current code, with this patch they will be ignored but not crash. I don't know that allowing mixed case env vars is wanted?
The only functional change here is to not crash if `is_dict` is called while there is mixed case env vars (starting with `AIRFLOW__` but then not all uppercase).

Comment (haard) [reply tests/core/test_configuration.py:89]: @ashb @XD-DENG I believe the PR behaviour is conservative (fixing the crash, not adding behaviour). Adding the inclusion of the MIXED_case env vars without breaking other things in the configuration module looks like it would take larger (in scope) changes.

Comment (ashb) [reply tests/core/test_configuration.py:89]: I'll take a look on Monday if XD doesn't get to it first

Comment (haard) [reply tests/core/test_configuration.py:89]: Ping @ashb @XD-DENG

File: airflow/executors/celery_executor.py
Comment (XD-DENG) [inline airflow/executors/celery_executor.py:119]: I assume the purpose of this change is to have trackback info in logs?

If there is other purpose(s) for this change, please clarify so we have full info. Thanks.

**Comment by XD-DENG** [comment]
Thanks @haard for the PR. I understand the only functional difference here is to avoid crash if there is "_mixed case env vars (starting with AIRFLOW__ but then not all uppercase)_". From that perspective, I agree this PR does its job and LGTM.

What we may want to discuss a bit further are different options:
- Option A: allow lowercases, and help users convert to upper case if necessary. This may not make sense given environment variables are case sensitive in Unix. But of course there may be different opinions.
- Option B: don't allow lowercase, and choose to ignore (what you are doing)
- Option C: don't allow lowercase, and reject explicitly. Maybe not making sense, because users may purposely want `AIRFLOW__CORE__lowercasekey` for whatever reason.

So overall, I agree with your direction, i.e. option B.

@ashb , may need your input here. Need your approval to unblock anyway :)

File: tests/core/test_configuration.py
Comment (XD-DENG) [reply tests/core/test_configuration.py:89]: I think what @ashb wants is actually aligned with what @haard is trying to do here.

We want to support the user-defined config sections/keys. But if user gives an env var like `AIRFLOW__S__key`, they expect to have something like `conf.as_dict().get('S').get('key')` OR `conf.as_dict().get('S').get('KEY')`. But actually the current behaviour is the whole process crashes.

What @haard does here is to avoid crash + enforce all-uppercase env var (otherwise ignored). Personally I'm ok with this.

File: airflow/executors/celery_executor.py
Comment (haard) [reply airflow/executors/celery_executor.py:119]: Yes, this is only to get the stack trace, since otherwise there is very little to help troubleshooting - I ran with this patch to be able to find the issue with mixed case env vars at all.

**Review by XD-DENG — APPROVED** [review]
Thanks again @haard . LGTM.

May you please rebase to the latest `master` branch before we can merge? Thanks

**Comment by XD-DENG** [comment]
The CI errors here are the same to the ones @kaxil has described in https://github.com/apache/airflow/pull/14694 .
Have validated that the tests relevant to this PR itself have passed. Gonna merge.

**Pull request merged**
